content/blog/2019-12-13-futil-fsm/index.md

```diff
@@ -0,0 +1,315 @@
+
+| FSM                  | Input Ports                                   | Output Ports                 |
+| -------------------- | --------------------------------------------- | ---------------------------- |
+| `enable`/`par`/`seq` | *val, rdy_A, rdy_B, ...*                      | *rdy, val_A, val_B, ...*     |
+| `if`                 | *val, rdy_con, cond, cond_read, rdy_T, rdy_F* | *rdy, val_con, val_T, val_F* |
+| `while`              | *val, rdy_con, cond, cond_read, rdy_body*     | *rdy, val_con, val_body*     |
+
+
+
+### Interfacing
+
+This pass creates input port *clock* for all components and *valid* for the top level component. Notice in FuTIL we actually does not have the notion of logical time step. However, to make things easier for RTL translation, we created these ports.
+
+#### MUX Signatures
+
+Similar to creating FSM signatures, we need to specify the inputs and outputs of each MUX component and the wires connecting each ports to its subcomponents. To do this, we create a Hashmap indexing with destination ports and store a vector of source ports according to the wiring of the component. For each destination port, if there are more than one source port connecting to it, we create an MUX.
+
+Notice there is a difference between control signal and data signals though. Control signals do not have corresponding *read* wires and no matter which control signal is high, the output is high. On the other hand, data signals always have corresponding *read* wires to explicitly specify if the data on the *data* wire is readable. The data signal should be chosen according to its *read* wires. Therefore at implementation, we go through wires twice. The first time we go through it we record *read* wires going to the same destination components. The second time we go through it we actually create the large MUX with both *data* and corresponding *read* wires.
+
+The last step is removing old wires connecting the same destination port with more than one source port.
+
+### Second Intermediate Pass
+
+#### FSM Implementation
+
+This pass creates true FSM representation in FuTIL AST.  Each `FSM` has a *name* field, a *states* Hashmap storing states and indexing of the state, a *start index* corresponding to the first state being created and a *last index* pointing to the last state being created. Each `State` is composed of a vector of *outputs*, where each *output* is specified with output *value* and *port name* and a vector for *transitions*, where each transition is a tuple of *next state index* and *inputs* of *value* and *port name*, so that the transition happens when the input has certain *value*. Finally, there is a default state for each state that is an optional field, telling which state it should transit to when no *transition* condition is met.
```

Comment:
- representation -> representations (plural)
- transit -> transition
- I am mildly surprised that a "default state" was necessary… I guess I would have thought it would make sense to just stay in the same state if no transition condition was met?

Reply:
This is the only thing we use the default state for. The reason this is there is because we were designing the FSM interface before we really knew what the FSMs would look like so we wanted to keep them general.